### PR TITLE
[android] Migrated MwmJobIntentService.java to MwmWork

### PR DESCRIPTION
**Pull request by sum20156**
Removed all classes using jobintentservice as it is deprecated, workmanager is used in the place.

Android O and above devices, it is not guaranteed that the jobIntentService starts immediately, but start with the delay depending on the device's condition. But, some phone manufacturers are deliberately canceling the job for better battery life . So the job is never called.
Workmanager try to execute the job immediately if it is not possible immediately it schedule it for later, but it assures it will be executed.

I have tested the flow, edits are uploading:
PHOTO-2023-03-04-22-01-41

Closes https://github.com/organicmaps/organicmaps/issues/4531

### CODE REVIEW

File: android/src/app/organicmaps/MwmWork.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public abstract class MwmWork extends Worker
+{
+
+  private final Context mContext;
+  private final WorkerParameters mWorkerParameters;
+  private static final String TAG = MwmWork.class.getSimpleName();
+
+  protected abstract void onHandleWorkInitialized(@NonNull WorkerParameters workerParameters);
+
+
```
Comment (biodranik) [inline android/src/app/organicmaps/MwmWork.java:21]:
```suggestion

```

File: android/src/app/organicmaps/MwmWork.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    super(context, workerParams);
+    this.mContext=context;
+    this.mWorkerParameters=workerParams;
+  }
+
+  @NonNull
+  @Override
+  public Result doWork()
+  {
+    MwmApplication app = MwmApplication.from(mContext);
```
Comment (biodranik) [inline android/src/app/organicmaps/MwmWork.java:39]:
```suggestion
    final MwmApplication app = MwmApplication.from(mContext);
```

File: android/src/app/organicmaps/MwmWork.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  @NonNull
+  protected String getTag()
+  {
+    return getClass().getSimpleName();
+  }
+
+  public MwmWork(@NonNull Context context, @NonNull WorkerParameters workerParams)
+  {
+    super(context, workerParams);
+    this.mContext=context;
```
Comment (biodranik) [inline android/src/app/organicmaps/MwmWork.java:31]: Did you read our style guide? 
```suggestion
    this.mContext = context;
```

File: android/src/app/organicmaps/MwmWork.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import android.content.Context;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.work.Worker;
+import androidx.work.WorkerParameters;
+import app.organicmaps.util.CrashlyticsUtils;
+import app.organicmaps.util.log.Logger;
+
+public abstract class MwmWork extends Worker
```
Comment (biodranik) [inline android/src/app/organicmaps/MwmWork.java:12]: Is that base class really needed?

File: android/src/app/organicmaps/MwmWork.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @NonNull
+  @Override
+  public Result doWork()
+  {
+    MwmApplication app = MwmApplication.from(mContext);
+    String msg = "onHandleWork: " + mWorkerParameters;
+    Logger.i(TAG, msg);
+    CrashlyticsUtils.INSTANCE.log(Log.INFO, getTag(), msg);
```
Comment (biodranik) [inline android/src/app/organicmaps/MwmWork.java:42]: Not needed.

**Review by biodranik — COMMENTED** [review]
This code can be made simpler.

How often is it run now?

File: android/src/app/organicmaps/MwmWork.java (hunk 4 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java:12]: currently it is not needed, in old code it was there, thats why I thought may be it is written to fullfill any longterm goal.

File: android/src/app/organicmaps/MwmWork.java (hunk 5 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java:42]: ok will remove it.

File: android/src/app/organicmaps/MwmWork.java (hunk 3 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java:31]: I had read it, apologies for forgot to to do code->reformat file

File: android/src/app/organicmaps/MwmWork.java (hunk 1 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java:21]: worker get executed everytime when we put app in the background

File: android/src/app/organicmaps/MwmWork.java (hunk 4 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java:12]: should I remove it?

File: android/src/app/organicmaps/background/OsmUploadWork.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    super(context, workerParams);
+  }
+
+  /**
+   * Starts this worker to upload map edits to osm servers.
+   */
+  public static void startActionUploadOsmChanges(@NonNull Context context)
+  {
+    WorkRequest workReq =new  OneTimeWorkRequest.Builder(OsmUploadWork.class)
+        .addTag(JobIdMap.getId(OsmUploadWork.class).toString()).build();
```
[the remark this answers is not in the document]
Comment (biodranik) [reply android/src/app/organicmaps/background/OsmUploadWork.java:28]: Is there a check that there are some OSM changes that await to be uploaded? It doesn't make any sense to start the background upload process if there is nothing there to upload.

**Comment by sum20156** [comment]
hi @rtsisyk , @biodranik done the necessary changes, please check and let me know if anything goes wrong.

**Comment by sum20156** [comment]
hi @biodranik , @rtsisyk any update?

File: android/src/app/organicmaps/MwmWork.java (hunk 5 above)
Comment (sum20156) [reply android/src/app/organicmaps/MwmWork.java]: i have removed it

File: android/src/app/organicmaps/background/OsmUploadWork.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sum20156) [reply android/src/app/organicmaps/background/OsmUploadWork.java]: I have added this check on `startActionUploadOsmChanges()` of `OsmUploadWork` so job will not be created

File: android/src/app/organicmaps/scheduling/JobIdMap.java [hunk 7]
```diff
@@ -10,7 +10,7 @@
   private static final Map<Class<?>, Integer> MAP = new HashMap<>();
 
   static {
-    MAP.put(OsmUploadService.class, calcIdentifier(MAP.size()));
+    MAP.put(OsmUploadWork.class, calcIdentifier(MAP.size()));
```
[the remark this answers is not in the document]
Comment (sum20156) [reply android/src/app/organicmaps/scheduling/JobIdMap.java]: removed

File: android/AndroidManifest.xml [hunk 8]
```diff
@@ -719,10 +719,7 @@
     <activity
       android:name="app.organicmaps.settings.DrivingOptionsActivity"
       android:label="@string/driving_options_title"/>
-    <service
-      android:name="app.organicmaps.background.OsmUploadService"
-      android:permission="android.permission.BIND_JOB_SERVICE"
-      android:exported="false"/>
+
```
Comment (biodranik) [inline android/AndroidManifest.xml:722]:
```suggestion
```

File: android/src/app/organicmaps/background/OsmUploadWork.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  /**
+   * Starts this worker to upload map edits to osm servers.
+   */
+  public static void startActionUploadOsmChanges(@NonNull Context context)
+  {
+    if (Editor.nativeHasSomethingToUpload() && OsmOAuth.isAuthorized(context))
+    {
+
```
Comment (biodranik) [inline android/src/app/organicmaps/background/OsmUploadWork.java:39]:
```suggestion
```

File: android/src/app/organicmaps/background/OsmUploadWork.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      final Constraints constraints = new Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED)
+                                                               .setRequiresCharging(true)
+                                                               .build();
+
```
Comment (biodranik) [inline android/src/app/organicmaps/background/OsmUploadWork.java:43]:
```suggestion
```

**Review by biodranik — APPROVED** [review]
LGTM if it works. @rtsisyk WDYT?

**Comment by rtsisyk** [comment]
After branching out the release, please.

**Comment by rtsisyk** [comment]
Up. Please finish with comments.

**Review by vng — APPROVED** [review]

**Comment by vng** [comment]
@sum20156 Unfortunately, some users claim that OSM edits are uploaded after 2-3 days. Even if the app is constantly used these days. For example on Android 10.
This is not good. Do we miss some settings for android.work.Worker here?

**Comment by sum20156** [comment]
> @sum20156 Unfortunately, some users claim that OSM edits are uploaded after 2-3 days. Even if the app is constantly used these days. For example on Android 10. This is not good. Do we miss some settings for android.work.Worker here?

no @vng I don't think so, the worker may not start immediately, but it also will not take 2-3 days to be triggered

**Comment by vng** [comment]
Ok, any idea why it may happen? Battery saving mode, bugs on vendor side, something else?
Can we do something in code? At least to make additional logging for better diagnostics.

**Comment by vng** [comment]
.setRequiresCharging(true) Do we need this?

**Comment by biodranik** [comment]
@vng You're right, we don't need it. Changes should be uploaded anytime (if mobile internet settings allow it).

**Pull request merged**